loader/loader.go

```diff
@@ -1188,110 +1188,246 @@ func fetchMatchedLiteral(ctx *tcontext.Context, router *router.Table, schema, ta
+				},
+			})
+		}
+		schemaFile := l.cfg.Dir + "/" + db + "-schema-create.sql" // cache friendly
+		err = dbRestoreQueue.push(&restoreSchemaJob{
+			loader:   l,
+			session:  dbSessionPool[dbSessionID],
+			database: db,
+			table:    "",
+			filepath: schemaFile,
+		})
+		if err != nil {
+			break
+		}
+	}
 
-		// create db
-		dbFile := fmt.Sprintf("%s/%s-schema-create.sql", l.cfg.Dir, db)
-		l.logger.Info("start to create schema", zap.String("schema file", dbFile))
-		err = l.restoreSchema(ctx, dbConn, dbFile, db)
+	// check producing error
+	if err != nil {
+		runtimeErr := dbRestoreQueue.close()
+		if errors.ErrorEqual(err, context.Canceled) {
+			err = runtimeErr
+		}
+		return err
+	}
+	// wait whole task done & close queue
+	err = dbRestoreQueue.close()
+	if err != nil {
+		return err
+	}
+
+	// run consumers of restore table schema queue
+	tblRestoreQueue := newJobQueue(ctx, concurrency, concurrency /** length of queue */)
+	tblRestoreQueue.startConsumers(func(ctx context.Context, job *restoreSchemaJob) error {
+		job.loader.logger.Info("start to create table", zap.String("table file", job.filepath))
+		err := job.loader.restoreTable(ctx, job.session, job.filepath, job.database, job.table)
 		if err != nil {
 			return err
 		}
-		l.logger.Info("finish to create schema", zap.String("schema file", dbFile))
+		job.loader.logger.Info("finish to create table", zap.String("table file", job.filepath))
+		return nil
+	})
 
-		tnames := make([]string, 0, len(tables))
-		for t := range tables {
-			tnames = append(tnames, t)
-		}
-		for _, table := range tnames {
-			dataFiles := tables[table]
-			tableFile := fmt.Sprintf("%s/%s.%s-schema.sql", l.cfg.Dir, db, table)
+	// push table schema restoring jobs to the queue
+tblSchemaLoop:
+	for dbSessionID, db := range dbs {
+		dbSessionID %= concurrency
+		for table := range l.db2Tables[db] {
+			schemaFile := l.cfg.Dir + "/" + db + "." + table + "-schema.sql" // cache friendly
 			if _, ok := l.tableInfos[tableName(db, table)]; !ok {
-				l.tableInfos[tableName(db, table)], err = parseTable(tctx, l.tableRouter, db, table, tableFile, l.cfg.LoaderConfig.SQLMode)
+				l.tableInfos[tableName(db, table)], err = parseTable(tctx, l.tableRouter, db, table, schemaFile, l.cfg.LoaderConfig.SQLMode)
 				if err != nil {
-					return terror.Annotatef(err, "parse table %s/%s", db, table)
+					err = terror.Annotatef(err, "parse table %s/%s", db, table)
+					break tblSchemaLoop
 				}
 			}
-
 			if l.checkPoint.IsTableFinished(db, table) {
 				l.logger.Info("table has finished, skip it.", zap.String("schema", db), zap.String("table", table))
 				continue
 			}
-
-			// create table
-			l.logger.Info("start to create table", zap.String("table file", tableFile))
-			err := l.restoreTable(ctx, dbConn, tableFile, db, table)
+			err = tblRestoreQueue.push(&restoreSchemaJob{
+				loader:   l,
+				session:  dbSessionPool[dbSessionID],
+				database: db,
+				table:    table,
+				filepath: schemaFile,
+			})
```

Comment:
It sames that there will be several jobs running at the same time and using the same mysql connection? If so, is there any problem with this situation?

Reply:
it's no problem for now, maybe use buffered channel implement  connection pool is more perfect?

Reply:
Yes. Or we can start connections for each consumer thread. And close them after we stop the consumers. Like we did in [dumpling](https://github.com/pingcap/dumpling/blob/master/v4/export/dump.go#L243).
Besides, `sql.Rows` will use `sql.Conn` exclusively. We'd better refine this part.

Reply:
> Besides, sql.Rows will use sql.Conn exclusively. We'd better refine this part.

I don't get it 🤣 , would you help me understand that? thx!

Reply:
> > Besides, sql.Rows will use sql.Conn exclusively. We'd better refine this part.

```go
	go func() {
		rows, _ := conn.QueryContext(ctx, "select * from pingcap.tb")
		for rows.Next() {
			time.Sleep(time.Minute)
		}
	}()
	time.Sleep(time.Second)
	rows, err := conn.QueryContext(ctx, "select * from pingcap.tb")
	if err != nil {
		return errors.Trace(err)
	}
	for rows.Next() {
        }
```

I mean this kind of code will cause an error.

```go
[mysql] 2021/03/26 18:20:32 packets.go:446: busy buffer
```

Reply:
> > > Besides, sql.Rows will use sql.Conn exclusively. We'd better refine this part.
> 
> ```go
> 	go func() {
> 		rows, _ := conn.QueryContext(ctx, "select * from pingcap.tb")
> 		for rows.Next() {
> 			time.Sleep(time.Minute)
> 		}
> 	}()
> 	time.Sleep(time.Second)
> 	rows, err := conn.QueryContext(ctx, "select * from pingcap.tb")
> 	if err != nil {
> 		return errors.Trace(err)
> 	}
> 	for rows.Next() {
>         }
> ```
> 
> I mean this kind of code will cause an error.
> 
> ```go
> [mysql] 2021/03/26 18:20:32 packets.go:446: busy buffer
> ```

Thanks a lot for your reminder! I don't even know that has such a deeply hidden problem. Fortunately, our modification will only perform DDL operations, so it will not trigger concurrent read buffer situations.

Reply:
this should has been fixed in https://github.com/pingcap/dm/pull/1466/commits/aae003edfc1dc3dc9e89a5f1f588833eee1ec252 @lichunzhu